ft_prepare_sourcemodel.m

```diff
@@ -725,6 +725,44 @@
   end
 end
 
+if strcmp(cfg.movetocentroid, 'yes')
+    % 1. compute centroids
+    % the FEM model should have tetraheders or hexaheders
+    centroids=[];
+    if isfield(headmodel, 'tet')
+      numtet = size(headmodel.tet, 1);
+      centroids.pos = zeros(numtet, 3);
+      for i=1:numtet
+        % compute the mean of the 4 corner points of the tetraheder
+        centroids.pos(i,:) = mean(headmodel.pos(headmodel.tet(i,:),:), 1);
+      end
+    elseif isfield(headmodel, 'hex')
+      numhex = size(headmodel.hex, 1);
+      centroids.pos = zeros(numhex, 3);
+      for i=1:numhex
+        % compute the mean of the 8 corner points of the hexaheder
+        centroids.pos(i,:) = mean(headmodel.pos(headmodel.hex(i,:),:), 1);
+      end
+    else
+      ft_error('the headmodel does not contain tetraheders or hexaheders');
+    end
+
+    % copy the specified fields, fields that are specified but not present will be silently ignored
+    centroids = copyfields(headmodel, centroids, {'tissue', 'tissuelabel', 'unit', 'coordsys'});
+
+    % 2. move the sourcemodel to the closest centroids
+    grid_shifted = zeros(size(sourcemodel.pos));
+    for i = 1:length(sourcemodel.pos)
+        [~, amin] = min(sum((sourcemodel.pos(i,:) - centroids.pos).^2,2));
+        grid_shifted(i,:) = centroids.pos(amin,:);
+    end
+    % eliminate duplicates (if, e.g., cfg.resolution is smaller than the
+    % mesh resolution)
+    sourcemodel.pos = unique(grid_shifted,'rows','stable');
```

Comment:
removing the duplicates might cause some applications of this to break: imagine starting with basedongrid or basedonresolution; in both cases `sourcemodel.dim` is used to determine how the data can be reshaped in a 3D array. After removing positions, the estimated data cannot be reshaped any more and `size(pos,1)` is not equal to `prod(dim)`. 

You either remove positions and remove the dim field, or you keep positions and dim. In the latter case, `ft_sourceplot` might still be able to plot it as if it were a regular 3D grid.

Reply:
I was about to write a comment on this (that is connected to setting `cfg.tight       = ft_getopt(cfg.sourcemodel, 'tight',       'no');` ).

I think `ft_sourceplot` works also with a sourcemodel that does not have the field "dim" (I need to check again). I would then rather eliminate the dim field in sourcemodel. 

If cfg.tight is set to yes,  then the dim is needed, but if the user wants a sourcemodel moved to the centroids, it does not seem necessary to have a smaller bounding box (i.e., to use cfg.tight), since it is already based on the volumetric mesh in input.

Maybe I missed something.

Reply:
The dim is present whenever the positions form a regular 3D grid, as in the example at the top of https://github.com/Donders-Institute/meg-hackathon/tree/master/2020-07-06-earth-moving-statfun. The tight option only applies when there is a dim. Please leave it as-is. 

The only thing you should worry about is whether you want to allow the slightly shifted positions to be still interpretable on their original regular positions (with a dim) or not (without a dim). Probably the best is not to keep the dim, since the positions are not exactly on a grid any more.

Reply:
It is not clear to me.
I removed the dim from the sourcemodel and removed the cfg.tight = no (inside the cfg.movetocentroids).
If I run:
```
cfg                 = [];
cfg.resolution      = 1; 
cfg.method          ='basedonresolution';
cfg.movetocentroids ='yes';
cfg.headmodel       = mesh;
cfg.headmodel.type  = 'simbio';
cfg.movetocentroid  = 'yes';
reg_grid = ft_prepare_sourcemodel(cfg);
```
it will enter cfg.tight (because is set to yes when choosing `cfg.method          ='basedonresolution';`) and will give an error because the field dim is not there.

I do not think you meant what I did.

Reply:
you should not specify both cfg.movetocentroid=yes and cfg.tight=yes. That is a user error, i.e. not something the code has to fix.

Reply:
it is the same as getting an error with cfg.movetocentroid=yes and giving it a spherical head model.

Reply:
I think still there is something to fix.
If the user specify `cfg.method          ='basedonresolution';`, then cfg.tight is automatically set to yes (without explicitly specifying it in the cfg). See line 221.
so, when I run:
```
cfg                 = [];
cfg.resolution      = 1; 
cfg.method          ='basedonresolution';
cfg.movetocentroids ='yes';
cfg.headmodel       = mesh;
cfg.headmodel.type  = 'simbio';
cfg.movetocentroid  = 'yes';
reg_grid = ft_prepare_sourcemodel(cfg);
```
is giving me the error:
Error using ft_prepare_sourcemodel (line 735) tightgrid only works for positions on a regular 3D grid

sorry if I am still missing something crucial.

Reply:
shall we remove line 221?